mmengine/dist/utils.py

```diff
@@ -169,16 +169,23 @@ def init_local_group(node_rank: int, num_gpus_per_node: int):
 
     Args:
         node_rank (int): Rank of machines used for training.
-        num_gpus_per_node (int): Number of gpus used for training in a single
-            machine.
+        num_gpus_per_node (list): Number of process used for training in all
+            machines.
     """  # noqa: W501
     global _LOCAL_PROCESS_GROUP
     assert _LOCAL_PROCESS_GROUP is None
 
-    ranks = list(
-        range(node_rank * num_gpus_per_node,
-              (node_rank + 1) * num_gpus_per_node))
-    _LOCAL_PROCESS_GROUP = torch_dist.new_group(ranks)
+    # `torch_dist.new_group` requires all process to enter the function call.
+    # `_LOCAL_PROCESS_GROUP` will be set as the process group of corresponding
+    # node.
+    num_machines = len(num_proc_per_node)
+    for i in range(num_machines):
+        start = sum(num_proc_per_node[:i]) if i != 0 else 0
+        end = sum(num_proc_per_node[:i + 1])
+        ranks_on_i = list(range(start, end))
```

Comment:
Are these lines the same as `ranks_on_i = num_proc_per_node[i]`?

Reply:
`ranks_on_i` should be a list of the global rank number of the current machine. `num_proc_per_node[i]` means the number of processes launched on this machine.